### PR TITLE
MBS-11348: New report about discIDs attached to a medium but not applied

### PR DESCRIPTION
**Pull request by loujine**
Implements MBS-11348

Find un-applied discIDs on releases where applying them should be a no-brainer: some track durations are unset on the release, and only discID is attached to the relevant medium.

# Checklist for author

PR ready to be merged, I think

# Action

Please check  lib/MusicBrainz/Server/Report/CDTOCReport.pm  which is used by both  CDTOCNotApplied.pm and CDTOCDubiousLength.pm

I modified generateHref() and entityHref() to be able to build urls with anchors, I think there are no side-effects but of course I might be wrong.
A far as I know urls with anchors are used to highlight track rows in releases (and it was probably never used with react), and I wanted to be able to highlight release rows on cdtoc pages too.
This last part is not required for the report and the two commits can be removed if needed.

### CODE REVIEW

File: root/report/components/CDTocReleaseList.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            <th>{l('Disc ID')}</th>
+            <th>{l('Release')}</th>
+            <th>{l('Artist')}</th>
+          </tr>
+        </thead>
+        <tbody>
+          {items.map((item, index) => {
+            return (
+              <tr className={loopParity(index)} key={item.cdtoc_id}>
+                {item.cdtoc ? (
```
Comment (yvanzo) [inline root/report/components/CDTocReleaseList.js:46]: Fix passing flow test by skipping rows where release doesn’t exist:

```suggestion
                {item.cdtoc && item.release ? (
```

File: root/static/scripts/common/components/CDTocLink.js [hunk 2]
```diff
@@ -15,10 +15,13 @@ type Props = {
   +cdToc: CDTocT,
   +content: string,
   +subPath?: string,
+  +anchorPath?: string,
 };
 
-const CDTocLink = ({cdToc, content, subPath}: Props): React.Element<'a'> => (
-  <a href={entityHref(cdToc, subPath)}>
+const CDTocLink = (
+  {cdToc, content, subPath, anchorPath}: Props
```
Comment (yvanzo) [inline root/static/scripts/common/components/CDTocLink.js:22]: Add missing comma using
```shell
git diff --name-only master '**.js' | xargs eslint --fix
```

```suggestion
  {cdToc, content, subPath, anchorPath}: Props,
```

File: root/report/components/CDTocReleaseList.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                  <>
+                    <td>
+                      <CDTocLink
+                        cdToc={item.cdtoc}
+                        content={item.cdtoc.discid}
+                        anchorPath={item.release.gid}
+                      />
```
Comment (yvanzo) [inline root/report/components/CDTocReleaseList.js:53]: Reorder props using
```shell
git diff --name-only master '**.js' | xargs ./script/check_unfixed_eslint_rules --fix
```

```suggestion
                      <CDTocLink
                        anchorPath={item.release.gid}
                        cdToc={item.cdtoc}
                        content={item.cdtoc.discid}
                      />
```

File: root/static/scripts/common/components/CDTocLink.js [hunk 4]
```diff
@@ -15,10 +15,13 @@ type Props = {
   +cdToc: CDTocT,
   +content: string,
   +subPath?: string,
+  +anchorPath?: string,
```
Comment (yvanzo) [inline root/static/scripts/common/components/CDTocLink.js:18]: Reorder props using the same
```shell
git diff --name-only master '**.js' | xargs ./script/check_unfixed_eslint_rules --fix
```

```suggestion
  +anchorPath?: string,
  +cdToc: CDTocT,
  +content: string,
  +subPath?: string,
```

**Review by yvanzo — CHANGES_REQUESTED** [review]
Thank you @loujine! Both reports work perfectly, and changes to support adding anchor are fine to me. Only one typing error and a few styling warnings to fix, see comments.

----

For the record, tested by running:
```shell
./admin/RunReports.pl CDTOCDubiousLength CDTOCNotApplied
```
And visiting the following pages:
* http://localhost:5000/reports
* http://localhost:5000/report/CDTOCDubiousLength
* http://localhost:5000/report/CDTOCNotApplied

Checked coding-style by running:
```shell
git diff --name-only master '**.js' | xargs ./node_modules/.bin/eslint
git diff --name-only master '**.js' | xargs ./script/check_unfixed_eslint_rules
git diff --name-only master '**.pm' | xargs perlcritic
```

File: root/report/components/CDTocReleaseList.js (hunk 1 above)
Comment (loujine) [reply root/report/components/CDTocReleaseList.js:46]: Completely missed that one, I will fix and amend

**Comment by loujine** [comment]
@yvanzo : thanks for the quick review, I think I fixed everything. Something is wrong with my eslint config, I have to find why I didn't get warnings

**Review by yvanzo — APPROVED** [review]

**Review by reosarevok — APPROVED** [review]
LGTM but did not test.

**Pull request merged**